include/siminusminus/containers/inmutableiterator.hpp

```diff
@@ -0,0 +1,167 @@
+#ifndef SIMINUSMINUS_CONTAINERS_INMUTABLEITERATOR_HPP
+#define SIMINUSMINUS_CONTAINERS_INMUTABLEITERATOR_HPP
+
+#include <algorithm>
+
+namespace cmm {
+namespace containers {
+
+/**
+ * \ingroup containers
+ * \brief iterator to inmutable objects.
+ * 
+ * The InmutableIterator class allows to create an iterator which cannot be
+ * modified, only allows read operations.
+ */
+template <typename T>
+class InmutableIterator : public std::iterator<std::input_iterator_tag, // iterator_category
+                                               const T* const 			// value_type
+                                              > 
+{
+
+public:
+
+	/**
+	 * Default constructor. Init the iterator to null.
+	 */
+	InmutableIterator(): _pointer(nullptr)
+	{
+
+	}
+
+	/**
+	 * Constructor that takes a T object to point to object memory direction.
+	 * @param object: object of T Type to init the iterator.
+	 */
+	InmutableIterator(const T& object) : _pointer(&object)
+	{
+		
+	}
+
+	/**
+	 * Copy constructor.
+	 * @param iit: InmutableIterator<T> object that we'll copy from.
+	 */
+	InmutableIterator(const InmutableIterator<T>& iit): InmutableIterator<T>(*iit._pointer) 
```

Comment:
Following with the comment above, this line triggers all my C++ pedant alarms: Dereferencing the iterator, that is, reading the element pointed to, should only happen if the user requests to, because an iterator is completely agnostic about what it is currently pointing to. If you do this, you might introduce some potential bugs (Specifically, this copy constructor segfaults in the following case:

``` cpp
InmutableIterator<int> it;
InmutableIterator<int> it2 = it; // Boom!
```

which is really annoying since the two lines above are perfectly fine).